# Actividad 1: Data Wrangling

**Objetivo:** Aplicar procesos de Data Wrangling dentro de una solución de Big Data.

**Alumno:** Martín Eduardo Chacón Orduño  
**Matrícula:** 351840  
**Materia:** Big Data  
**Actividad:** 3.6 Actividad 1  
**Fecha:** 30/09/2026

---

En este notebook se documentan diferentes operaciones de Data Wrangling
utilizando Python, pandas y el conjunto de datos Superstore.

In [315]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

print(f"Numpy: {np.__version__}")
print(f"pandas: {pd.__version__}")

Numpy: 2.5.3
pandas: 3.0.6


In [316]:
file = Path("Superstore.xls")

print(f"Notebook directory: {Path.cwd()}")
print(f"Does the file exists? {file.exists()}")

Notebook directory: /Users/martinchacon/Documents/semestre-2026-b/big-data/actividades/3.6-actividad-1
Does the file exists? True


## Ejercicio 4.01: Examinar los datos de ventas de Superstore

En este ejercicio se carga la hoja `Orders` del archivo `Superstore.xls`.
Después se examinan los primeros registros, las columnas y las dimensiones del
conjunto de datos. Finalmente, se elimina la columna `Row ID` porque pandas ya
proporciona un índice para identificar las filas.

In [317]:
df = pd.read_excel(file, sheet_name="Orders")
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [318]:
print(f"Nomber of rows: {df.shape[0]:,}")
print(f"Number of columns: {df.shape[1]:,}")

print("\nAvailable columns:")
for column in df.columns:
    print(f"-{column}")

Nomber of rows: 9,994
Number of columns: 21

Available columns:
-Row ID
-Order ID
-Order Date
-Ship Date
-Ship Mode
-Customer ID
-Customer Name
-Segment
-Country
-City
-State
-Postal Code
-Region
-Product ID
-Category
-Sub-Category
-Product Name
-Sales
-Quantity
-Discount
-Profit


###Eliminación de Row ID
La columna `Row ID` contiene un identificador consecutivo que resulta redundante,
ya que el DataFrame posee su propio índice. Se elimina la columna 
y se vuelven a comprobar sus dimensiones

In [319]:
df.drop(columns="Row ID", inplace=True)
df.head()

print(f"Number of rows after the drop: {df.shape[0]:,}")
print(f"Number of columns after the drop: {df.shape[1]:,}")
print(f"Final dimensions: {df.shape}")

Number of rows after the drop: 9,994
Number of columns after the drop: 20
Final dimensions: (9994, 20)


### Resultado del ejercicio 4.01
La hoja `Orders` contenia inicialmente 9,994 registros y 21 columnas. Después de
eliminar `Row ID`, el DataFrame conservó los 9,994 registros y quedó con
20 columnas útiles.

## Ejercicio 4.02: Función `Unique`

La función `unique()` permite obtener los valores diferente sd una columna, mientras que `nunique()` permite contar cuántos valores distintos existen.

Estas funciones se utilizarán para examinar las columnas `State`, `Country` y `City` 
del conjunto de datos.

In [320]:
unqiue_states = df["State"].unique()

print(f"Present states in data:")
print(unqiue_states)

Present states in data:
<StringArray>
[            'Kentucky',           'California',              'Florida',
       'North Carolina',           'Washington',                'Texas',
            'Wisconsin',                 'Utah',             'Nebraska',
         'Pennsylvania',             'Illinois',            'Minnesota',
             'Michigan',             'Delaware',              'Indiana',
             'New York',              'Arizona',             'Virginia',
            'Tennessee',              'Alabama',       'South Carolina',
               'Oregon',             'Colorado',                 'Iowa',
                 'Ohio',             'Missouri',             'Oklahoma',
           'New Mexico',            'Louisiana',          'Connecticut',
           'New Jersey',        'Massachusetts',              'Georgia',
               'Nevada',         'Rhode Island',          'Mississippi',
             'Arkansas',              'Montana',        'New Hampshire',
             

In [321]:
number_of_unique_states = df["State"].nunique()

print(f"Quantity of unique states: {number_of_unique_states}")

Quantity of unique states: 49


In [322]:
unqiue_countrys = df["Country"].unique()

print(f"Found countrys: {unqiue_countrys}")

Found countrys: <StringArray>
['United States']
Length: 1, dtype: str


In [323]:
citys_number = df["City"].nunique()
print(f"Quantity of unique cities: {citys_number}")

Quantity of unique cities: 531


### Resultado del ejercicio 4.02

El conjunto de datos contiene información de 49 estados y 531 ciudades. La
columna `Country` tiene un solo valor único: `United States`.

Esto indica que `State` y `City` aportan variabilidad al análisis. En cambio,
`Country` no permite comparar diferentes países y podría eliminarse si no fuera
necesaria en análisis posteriores.

## Ejercicio 4.03: Establecer y restablecer el índice

En este ejercicio se crea un DataFrame pequeño para practicar las operaciones
relacionadas con el índice. Se utilizarán etiquetas personalizadas, se
restablecerá el índice y se convertirá una columna en el nuevo índice.

In [324]:
matrix_data = np.matrix(
    "22,66,140;"
    "42,70,148;"
    "30,62,125;"
    "35,68,160;"
    "25,62,152"
)

row_labels = ["A", "B", "C", "D", "E"]
column_headings = ["Age", "Height", "Weight"]

print("Data matrix:")
print(matrix_data)

Data matrix:
[[ 22  66 140]
 [ 42  70 148]
 [ 30  62 125]
 [ 35  68 160]
 [ 25  62 152]]


In [325]:
df1 = pd.DataFrame(data=matrix_data, index=row_labels, columns=column_headings)
df1

,Age,Height,Weight
A,22,66,140
B,42,70,148
C,30,62,125
D,35,68,160
E,25,62,152


### Restablecimiento del índice

El método `reset_index()` reemplaza las etiquetas actuales por un índice
numérico. De manera predeterminada, las etiquetas anteriores se conservan en
una nueva columna llamada `index`.

In [326]:
df1_reset = df1.reset_index()
df1_reset

,index,Age,Height,Weight
0,A,22,66,140
1,B,42,70,148
2,C,30,62,125
3,D,35,68,160
4,E,25,62,152


Cuando se utiliza `drop=True`, el índice se restablece, pero las etiquetas
anteriores no se agregan como una nueva columna.

In [327]:
df1_reset_drop =df1.reset_index(drop=True)
df1_reset_drop

,Age,Height,Weight
0,22,66,140
1,42,70,148
2,30,62,125
3,35,68,160
4,25,62,152


### Creación de un nuevo índice

Se agrega la columna `Profession`, que contiene una profesión para cada
registro. Posteriormente, esta columna se utilizará como índice.

In [328]:
df1["Profession"] = [
    "Student",
    "Teacher",
    "Engineer",
    "Doctor",
    "Nurse"
]
df1

df1_profession = df1.set_index("Profession")

df1_profession

print("Nombre del índice:", df1_profession.index.name)
print("Valores del índice:", df1_profession.index.tolist())

Nombre del índice: Profession
Valores del índice: ['Student', 'Teacher', 'Engineer', 'Doctor', 'Nurse']


## Ejercicio 4.04: Método GroupBy

El método `groupby()` permite agrupar registros que comparten una característica. 
Después de formar los grupos, se pueden aplicar operaciones estadísticas como 
la media, la suma o la descripción de los datos.

In [329]:
df_subset = df.loc[
    range(10),
    ["Ship Mode", "State", "Sales"]
].copy()

df_subset

,Ship Mode,State,Sales
0,Second Class,Kentucky,261.9600
1,Second Class,Kentucky,731.9400
2,Second Class,California,14.6200
3,Standard Class,Florida,957.5775
4,Standard Class,Florida,22.3680
5,Standard Class,California,48.8600
6,Standard Class,California,7.2800
7,Standard Class,California,907.1520
8,Standard Class,California,18.5040
9,Standard Class,California,114.9000


In [330]:
per_state = df_subset.groupby("State")
per_state

In [331]:
average_sales_state = (per_state["Sales"].mean().to_frame(name="Average Sales"))
average_sales_state

,Average Sales
State,
California,185.219333
Florida,489.972750
Kentucky,496.950000


In [332]:
total_sales_stae = (per_state["Sales"].sum().to_frame(name="Total sales"))
total_sales_stae


,Total sales
State,
California,1111.3160
Florida,979.9455
Kentucky,993.9000


In [333]:
california_statistics = pd.DataFrame(per_state["Sales"].describe().loc["California"])
california_statistics.columns = ["California"]
california_statistics

,California
count,6.000000
mean,185.219333
std,355.889307
min,7.280000
25%,15.591000
50%,33.682000
75%,98.390000
max,907.152000


In [334]:
ship_statistics = (
    df_subset
    .groupby("Ship Mode")["Sales"]
    .describe()
    .loc[["Second Class", "Standard Class"]]
)

ship_statistics

,count,mean,std,min,25%,50%,75%,max
Ship Mode,,,,,,,,
Second Class,3.0,336.173333,364.373037,14.62,138.290,261.96,496.950,731.9400
Standard Class,7.0,296.663071,435.947552,7.28,20.436,48.86,511.026,957.5775


In [335]:
city_state = (
    df
    .groupby(["State", "City"])["Sales"]
    .describe()
)

city_state.head(10)

count        mean         std     min      25%  \
State   City                                                            
Alabama Auburn           6.0  294.471667  361.914543   3.760   8.8050   
        Decatur         13.0  259.601538  385.660903  14.940  23.9200   
        Florence         5.0  399.470000  796.488863   4.980   7.2700   
        Hoover           4.0  131.462500  230.646923   7.160  13.3925   
        Huntsville      10.0  248.437000  419.576667   3.620  26.8700   
        Mobile          11.0  496.635455  914.087425   8.960  46.8600   
        Montgomery      10.0  372.273000  475.397645  10.160  21.7075   
        Tuscaloosa       2.0   87.850000   76.523096  33.740  60.7950   
Arizona Avondale         6.0  157.801333  288.247527  14.576  18.1480   
        Bullhead City    2.0   11.144000    4.559425   7.920   9.5320   

                           50%       75%       max  
State   City                                        
Alabama Auburn         182.030  456.4075   900.080  
        Decatur         44.950  239.9200  1215.920  
        Florence        12.480  152.7600  1819.860  
        Hoover          20.725  138.7950   477.240  
        Huntsville      81.920  171.8075  1319.960  
        Mobile          70.980  505.9650  3040.000  
        Montgomery     187.215  499.0550  1394.950  
        Tuscaloosa      87.850  114.9050   141.960  
Arizona Avondale        35.596   88.6780   742.336  
        Bullhead City   11.144   12.7560    14.368

## Ejercicio 4.05: Tratamiento de valores faltantes

En este ejercicio se aplican diferentes técnicas para completar valores faltantes:

- Sustitución mediante un valor fijo.
- Relleno con el valor anterior.
- Relleno con el valor siguiente.
- Sustitución mediante el promedio.

In [336]:
df_missing = pd.read_excel(file, sheet_name="Missing")
df_missing

,Customer,Product,Sales,Quantity,Discount,Profit
0,Brosina Hoffman,NaN,1706.184,9.0,0.2,85.3092
1,Brosina Hoffman,Phones,911.424,4.0,0.2,68.3568
2,Zuschuss Donatelli,Art,8.560,2.0,0.0,2.4824
3,Zuschuss Donatelli,Phones,NaN,3.0,0.2,16.0110
4,Zuschuss Donatelli,Binders,22.720,4.0,0.2,7.3840
5,Eric Hoffmann,Binders,11.648,NaN,0.2,4.2224
6,Eric Hoffmann,Accessories,90.570,3.0,0.0,11.7741
7,Ruben Ausman,NaN,77.880,2.0,0.0,NaN
8,NaN,Accessories,13.980,2.0,0.0,6.1512
9,Kunst Miller,Binders,25.824,6.0,0.2,9.3612


In [337]:
missing_values = df_missing.isna().sum()
missing_values

Customer    1
Product     2
Sales       1
Quantity    1
Discount    0
Profit      1
dtype: int64

In [338]:
total_fill = df_missing.fillna("FILL")
total_fill

,Customer,Product,Sales,Quantity,Discount,Profit
0,Brosina Hoffman,FILL,1706.184,9.0,0.2,85.3092
1,Brosina Hoffman,Phones,911.424,4.0,0.2,68.3568
2,Zuschuss Donatelli,Art,8.56,2.0,0.0,2.4824
3,Zuschuss Donatelli,Phones,FILL,3.0,0.2,16.011
4,Zuschuss Donatelli,Binders,22.72,4.0,0.2,7.384
5,Eric Hoffmann,Binders,11.648,FILL,0.2,4.2224
6,Eric Hoffmann,Accessories,90.57,3.0,0.0,11.7741
7,Ruben Ausman,FILL,77.88,2.0,0.0,FILL
8,FILL,Accessories,13.98,2.0,0.0,6.1512
9,Kunst Miller,Binders,25.824,6.0,0.2,9.3612


In [339]:
fill_text = (df_missing[["Customer", "Product"]].fillna("FILL"))
fill_text


,Customer,Product
0,Brosina Hoffman,FILL
1,Brosina Hoffman,Phones
2,Zuschuss Donatelli,Art
3,Zuschuss Donatelli,Phones
4,Zuschuss Donatelli,Binders
5,Eric Hoffmann,Binders
6,Eric Hoffmann,Accessories
7,Ruben Ausman,FILL
8,FILL,Accessories
9,Kunst Miller,Binders


In [340]:
fill_sales = (df_missing["Sales"].ffill())
fill_sales

0     1706.184
1      911.424
2        8.560
3        8.560
4       22.720
5       11.648
6       90.570
7       77.880
8       13.980
9       25.824
10     146.730
Name: Sales, dtype: float64

In [341]:
bfill_sales = df_missing["Sales"].bfill()
bfill_sales

0     1706.184
1      911.424
2        8.560
3       22.720
4       22.720
5       11.648
6       90.570
7       77.880
8       13.980
9       25.824
10     146.730
Name: Sales, dtype: float64

In [342]:
average_sales = df_missing["Sales"].mean()
print(f"Average sales: {average_sales:.3f}")

Average sales: 301.552


In [343]:
sales_with_average = df_missing["Sales"].fillna(average_sales)
sales_with_average

0     1706.184
1      911.424
2        8.560
3      301.552
4       22.720
5       11.648
6       90.570
7       77.880
8       13.980
9       25.824
10     146.730
Name: Sales, dtype: float64

## Ejercicio 4.06: Eliminación de valores faltantes con dropna()

El método `dropna()` permite eliminar filas o columnas que contienen valores 
faltantes. El parámetro `axis` determina si se eliminan filas o columnas, mientras 
que `thresh` establece la cantidad mínima de valores válidos que deben conservarse.

In [344]:
print(f"Original dimensions: {df_missing.shape}")
df_missing

Original dimensions: (11, 6)


,Customer,Product,Sales,Quantity,Discount,Profit
0,Brosina Hoffman,NaN,1706.184,9.0,0.2,85.3092
1,Brosina Hoffman,Phones,911.424,4.0,0.2,68.3568
2,Zuschuss Donatelli,Art,8.560,2.0,0.0,2.4824
3,Zuschuss Donatelli,Phones,NaN,3.0,0.2,16.0110
4,Zuschuss Donatelli,Binders,22.720,4.0,0.2,7.3840
5,Eric Hoffmann,Binders,11.648,NaN,0.2,4.2224
6,Eric Hoffmann,Accessories,90.570,3.0,0.0,11.7741
7,Ruben Ausman,NaN,77.880,2.0,0.0,NaN
8,NaN,Accessories,13.980,2.0,0.0,6.1512
9,Kunst Miller,Binders,25.824,6.0,0.2,9.3612


In [345]:
no_incomplete_rows =df_missing.dropna(axis=0)

print(f"Dimensions: {no_incomplete_rows.shape}")
no_incomplete_rows

Dimensions: (6, 6)


,Customer,Product,Sales,Quantity,Discount,Profit
1,Brosina Hoffman,Phones,911.424,4.0,0.2,68.3568
2,Zuschuss Donatelli,Art,8.560,2.0,0.0,2.4824
4,Zuschuss Donatelli,Binders,22.720,4.0,0.2,7.3840
6,Eric Hoffmann,Accessories,90.570,3.0,0.0,11.7741
9,Kunst Miller,Binders,25.824,6.0,0.2,9.3612
10,Kunst Miller,Paper,146.730,3.0,0.0,68.9631


In [346]:
no_incomplete_columns = df_missing.dropna(axis=1)

print(f"Dimensios: {no_incomplete_columns.shape}")
no_incomplete_columns

Dimensios: (11, 1)


,Discount
0,0.2
1,0.2
2,0.0
3,0.2
4,0.2
5,0.2
6,0.0
7,0.0
8,0.0
9,0.2


In [347]:
columns_thresh = df_missing.dropna(axis=1, thresh=10)
columns_thresh

,Customer,Sales,Quantity,Discount,Profit
0,Brosina Hoffman,1706.184,9.0,0.2,85.3092
1,Brosina Hoffman,911.424,4.0,0.2,68.3568
2,Zuschuss Donatelli,8.560,2.0,0.0,2.4824
3,Zuschuss Donatelli,NaN,3.0,0.2,16.0110
4,Zuschuss Donatelli,22.720,4.0,0.2,7.3840
5,Eric Hoffmann,11.648,NaN,0.2,4.2224
6,Eric Hoffmann,90.570,3.0,0.0,11.7741
7,Ruben Ausman,77.880,2.0,0.0,NaN
8,NaN,13.980,2.0,0.0,6.1512
9,Kunst Miller,25.824,6.0,0.2,9.3612


### Interpretación

- `axis=0` elimina las filas que contienen uno o más valores faltantes.
- `axis=1` elimina las columnas que contienen uno o más valores faltantes.
- `thresh=10` conserva las columnas que tienen al menos 10 valores válidos.
- Las operaciones se guardaron en nuevos DataFrames, por lo que `df_missing`
  no fue modificado.

## Ejercicio 4.07: Concatenación de conjuntos de datos

La función `pd.concat()` permite combinar varios DataFrames. La concatenación 
puede realizarse verticalmente, agregando filas, u horizontalmente, agregando 
columnas.

El parámetro `axis` determina la dirección de la concatenación:

- `axis=0`: combina los DataFrames por filas.
- `axis=1`: combina los DataFrames por columnas.

In [348]:
print(f"Original df dimensions: {df.shape}")

df.head()

Original df dimensions: (9994, 20)


,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [349]:
from IPython.core.display_functions import display
sample_columns = ["Customer Name", "State", "Sales", "Profit"]

df_1 = df[sample_columns].sample(n=4, random_state=1)
df_2 = df[sample_columns].sample(n=4, random_state=2)
df_3 = df[sample_columns].sample(n=4, random_state=3)

print("Sample 1:")
display(df_1)

print("Sample 2:")
display(df_2)

print("Sample 3:")
display(df_3)

Sample 1:


,Customer Name,State,Sales,Profit
1814,Stefania Perrino,California,170.136,-8.5068
881,Dave Kipp,Virginia,30.840,8.3268
1122,Hunter Lopez,Arkansas,22.960,6.6584
6807,Ed Braxton,Texas,98.376,35.6613


Sample 2:


,Customer Name,State,Sales,Profit
4607,Maxwell Schwartz,New York,535.410,160.6230
9106,Adrian Barton,Texas,287.910,33.5895
5481,Chris Selesnick,California,36.000,6.4800
4506,Ben Ferrer,California,818.376,51.1485


Sample 3:


,Customer Name,State,Sales,Profit
3691,Frank Atkinson,Colorado,439.992,164.9970
6544,Darren Powers,Ohio,58.170,-46.5360
5972,Bryan Davis,New York,240.784,30.0980
7968,Corey Catlett,Colorado,15.992,-2.9985


In [350]:
df_cat1 = pd.concat([df_1, df_2, df_3], axis=0)
print(f"Dimensions: {df_cat1.shape}")
df_cat1

Dimensions: (12, 4)


,Customer Name,State,Sales,Profit
1814,Stefania Perrino,California,170.136,-8.5068
881,Dave Kipp,Virginia,30.840,8.3268
1122,Hunter Lopez,Arkansas,22.960,6.6584
6807,Ed Braxton,Texas,98.376,35.6613
4607,Maxwell Schwartz,New York,535.410,160.6230
9106,Adrian Barton,Texas,287.910,33.5895
5481,Chris Selesnick,California,36.000,6.4800
4506,Ben Ferrer,California,818.376,51.1485
3691,Frank Atkinson,Colorado,439.992,164.9970
6544,Darren Powers,Ohio,58.170,-46.5360


In [351]:
df_cat1_new_indexes = pd.concat([df_1, df_2, df_3], axis=0, ignore_index=True)
df_cat1_new_indexes

,Customer Name,State,Sales,Profit
0,Stefania Perrino,California,170.136,-8.5068
1,Dave Kipp,Virginia,30.840,8.3268
2,Hunter Lopez,Arkansas,22.960,6.6584
3,Ed Braxton,Texas,98.376,35.6613
4,Maxwell Schwartz,New York,535.410,160.6230
5,Adrian Barton,Texas,287.910,33.5895
6,Chris Selesnick,California,36.000,6.4800
7,Ben Ferrer,California,818.376,51.1485
8,Frank Atkinson,Colorado,439.992,164.9970
9,Darren Powers,Ohio,58.170,-46.5360


In [352]:
df_cat1 = pd.concat([df_1, df_2, df_3], axis=1)
print(f"Dimensions: {df_cat1.shape}")
df_cat1

Dimensions: (12, 12)


,Customer Name,State,Sales,Profit,Customer Name,State,Sales,Profit,Customer Name,State,Sales,Profit
1814,Stefania Perrino,California,170.136,-8.5068,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
881,Dave Kipp,Virginia,30.840,8.3268,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1122,Hunter Lopez,Arkansas,22.960,6.6584,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6807,Ed Braxton,Texas,98.376,35.6613,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4607,NaN,NaN,NaN,NaN,Maxwell Schwartz,New York,535.410,160.6230,NaN,NaN,NaN,NaN
9106,NaN,NaN,NaN,NaN,Adrian Barton,Texas,287.910,33.5895,NaN,NaN,NaN,NaN
5481,NaN,NaN,NaN,NaN,Chris Selesnick,California,36.000,6.4800,NaN,NaN,NaN,NaN
4506,NaN,NaN,NaN,NaN,Ben Ferrer,California,818.376,51.1485,NaN,NaN,NaN,NaN
3691,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Frank Atkinson,Colorado,439.992,164.9970
6544,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Darren Powers,Ohio,58.170,-46.5360


### Conclusión

La función `pd.concat()` permitió combinar tres muestras del conjunto de datos. 
Al utilizar `axis=0`, las muestras se agregaron verticalmente y se obtuvo un 
DataFrame de 12 filas y 4 columnas.

Al utilizar `axis=1`, las muestras se combinaron horizontalmente. Debido a que 
cada muestra conservaba índices diferentes, pandas alineó los registros mediante 
esos índices y colocó valores `NaN` donde no existía una coincidencia.

La concatenación vertical resulta útil para incorporar nuevos registros, mientras 
que la concatenación horizontal puede utilizarse para agregar nuevas características 
cuando los DataFrames comparten los mismos índices.

## Ejercicio 4.08: Combinación mediante una clave común

La función `pd.merge()` permite combinar DataFrames mediante una columna que 
funciona como clave común.

En este ejercicio se utilizan dos tipos de combinación:

- `inner`: conserva solamente las claves que existen en ambos DataFrames.
- `outer`: conserva todas las claves, aunque no tengan una coincidencia.

In [353]:
df_1 = (df[["Ship Date", "Ship Mode", "Customer Name"]].iloc[0:4].copy())
df_1

,Ship Date,Ship Mode,Customer Name
0,2016-11-11,Second Class,Claire Gute
1,2016-11-11,Second Class,Claire Gute
2,2016-06-16,Second Class,Darrin Van Huff
3,2015-10-18,Standard Class,Sean O'Donnell


In [354]:
df_2 = (df[["Customer Name", "Product Name", "Quantity"]].iloc[0:4].copy())
df_2

,Customer Name,Product Name,Quantity
0,Claire Gute,Bush Somerset Collection Bookcase,2
1,Claire Gute,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",3
2,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
3,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5


In [355]:
inside_union = pd.merge(df_1, df_2, on="Customer Name", how="inner")
print(f"Dimensions: {inside_union.shape}")
inside_union

Dimensions: (6, 5)


,Ship Date,Ship Mode,Customer Name,Product Name,Quantity
0,2016-11-11,Second Class,Claire Gute,Bush Somerset Collection Bookcase,2
1,2016-11-11,Second Class,Claire Gute,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",3
2,2016-11-11,Second Class,Claire Gute,Bush Somerset Collection Bookcase,2
3,2016-11-11,Second Class,Claire Gute,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",3
4,2016-06-16,Second Class,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
5,2015-10-18,Standard Class,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5


In [356]:
no_duplicates_inside_union = (inside_union.drop_duplicates().reset_index(drop=True))
print(f"Dimensions: {no_duplicates_inside_union.shape}")
no_duplicates_inside_union

Dimensions: (4, 5)


,Ship Date,Ship Mode,Customer Name,Product Name,Quantity
0,2016-11-11,Second Class,Claire Gute,Bush Somerset Collection Bookcase,2
1,2016-11-11,Second Class,Claire Gute,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",3
2,2016-06-16,Second Class,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
3,2015-10-18,Standard Class,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5


In [357]:
df_3 = (df[["Customer Name", "Product Name", "Quantity"]].iloc[2:6].copy())
df_3

,Customer Name,Product Name,Quantity
2,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
3,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5
4,Sean O'Donnell,Eldon Fold 'N Roll Cart System,2
5,Brosina Hoffman,Eldon Expressions Wood and Plastic Desk Access...,7


In [358]:
union_inner_13 = (pd.merge(df_1, df_3, on="Customer Name", how="inner").drop_duplicates().reset_index(drop=True))
print(f"Dimensions: {union_inner_13.shape}")
union_inner_13

Dimensions: (3, 5)


,Ship Date,Ship Mode,Customer Name,Product Name,Quantity
0,2016-06-16,Second Class,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
1,2015-10-18,Standard Class,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5
2,2015-10-18,Standard Class,Sean O'Donnell,Eldon Fold 'N Roll Cart System,2


In [359]:
union_outer_13 = (pd.merge(df_1, df_3, on="Customer Name", how="outer").drop_duplicates().reset_index(drop=True))
print(f"Dimensions: {union_outer_13.shape}")
union_outer_13

Dimensions: (5, 5)


,Ship Date,Ship Mode,Customer Name,Product Name,Quantity
0,NaT,NaN,Brosina Hoffman,Eldon Expressions Wood and Plastic Desk Access...,7.0
1,2016-11-11,Second Class,Claire Gute,NaN,NaN
2,2016-06-16,Second Class,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2.0
3,2015-10-18,Standard Class,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5.0
4,2015-10-18,Standard Class,Sean O'Donnell,Eldon Fold 'N Roll Cart System,2.0


In [360]:
origin_union = (
    pd.merge(
        df_1,
        df_3,
        on="Customer Name",
        how="outer",
        indicator=True
    )
    .drop_duplicates()
    .reset_index(drop=True)
)

origin_union

,Ship Date,Ship Mode,Customer Name,Product Name,Quantity,_merge
0,NaT,NaN,Brosina Hoffman,Eldon Expressions Wood and Plastic Desk Access...,7.0,right_only
1,2016-11-11,Second Class,Claire Gute,NaN,NaN,left_only
2,2016-06-16,Second Class,Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2.0,both
3,2015-10-18,Standard Class,Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5.0,both
4,2015-10-18,Standard Class,Sean O'Donnell,Eldon Fold 'N Roll Cart System,2.0,both


### Interpretación

La combinación interna conservó solamente los clientes presentes en ambos 
DataFrames. Por esta razón, al combinar `df_1` y `df_3`, se conservaron los 
registros de Darrin Van Huff y Sean O'Donnell.

La combinación externa conservó todos los clientes. Cuando un cliente no tenía 
información correspondiente en la otra tabla, pandas agregó valores `NaN` o `NaT`.

### Conclusión

La función `pd.merge()` permite relacionar diferentes tablas mediante una clave 
común. La combinación `inner` resulta útil cuando solamente interesan los registros 
coincidentes, mientras que la combinación `outer` permite conservar toda la 
información disponible.

También se observó que una clave repetida puede generar varias combinaciones. 
Por ello, es importante revisar y eliminar duplicados cuando el contexto del análisis 
así lo requiera.

## Ejercicio 4.09: Método join()

El método `join()` combina las columnas de dos DataFrames utilizando sus índices. 
En este ejercicio, la columna `Customer Name` se convierte en el índice de ambas 
tablas y se aplican combinaciones `left`, `right`, `inner` y `outer`.

In [361]:
df_1 = df[["Customer Name", "Ship Date", "Ship Mode"]].iloc[0:4].copy().set_index("Customer Name")
df_1

,Ship Date,Ship Mode
Customer Name,,
Claire Gute,2016-11-11,Second Class
Claire Gute,2016-11-11,Second Class
Darrin Van Huff,2016-06-16,Second Class
Sean O'Donnell,2015-10-18,Standard Class


In [362]:
df_2 = df[["Customer Name", "Product Name", "Quantity"]].iloc[2:6].copy().set_index("Customer Name")
df_2

,Product Name,Quantity
Customer Name,,
Darrin Van Huff,Self-Adhesive Address Labels for Typewriters b...,2
Sean O'Donnell,Bretford CR4500 Series Slim Rectangular Table,5
Sean O'Donnell,Eldon Fold 'N Roll Cart System,2
Brosina Hoffman,Eldon Expressions Wood and Plastic Desk Access...,7


In [363]:
print("f_1 index:")
print(df_1.index.tolist())

print("\ndf_2 index:")
print(df_2.index.tolist())

f_1 index:
['Claire Gute', 'Claire Gute', 'Darrin Van Huff', "Sean O'Donnell"]

df_2 index:
['Darrin Van Huff', "Sean O'Donnell", "Sean O'Donnell", 'Brosina Hoffman']


In [364]:
left_join = df_1.join(df_2, how="left").drop_duplicates()

print(f"Dimensions: {left_join.shape}")
left_join

Dimensions: (4, 4)


,Ship Date,Ship Mode,Product Name,Quantity
Customer Name,,,,
Claire Gute,2016-11-11,Second Class,NaN,NaN
Darrin Van Huff,2016-06-16,Second Class,Self-Adhesive Address Labels for Typewriters b...,2.0
Sean O'Donnell,2015-10-18,Standard Class,Bretford CR4500 Series Slim Rectangular Table,5.0
Sean O'Donnell,2015-10-18,Standard Class,Eldon Fold 'N Roll Cart System,2.0


In [365]:
right_join = df_1.join(df_2, how="right").drop_duplicates()

print(f"Dimensions: {right_join.shape}")
right_join

Dimensions: (4, 4)


,Ship Date,Ship Mode,Product Name,Quantity
Customer Name,,,,
Darrin Van Huff,2016-06-16,Second Class,Self-Adhesive Address Labels for Typewriters b...,2
Sean O'Donnell,2015-10-18,Standard Class,Bretford CR4500 Series Slim Rectangular Table,5
Sean O'Donnell,2015-10-18,Standard Class,Eldon Fold 'N Roll Cart System,2
Brosina Hoffman,NaT,NaN,Eldon Expressions Wood and Plastic Desk Access...,7


In [366]:
inner_join = df_1.join(df_2, how="inner").drop_duplicates()

print(f"Dimensions: {inner_join.shape}")
inner_join

Dimensions: (3, 4)


,Ship Date,Ship Mode,Product Name,Quantity
Customer Name,,,,
Darrin Van Huff,2016-06-16,Second Class,Self-Adhesive Address Labels for Typewriters b...,2
Sean O'Donnell,2015-10-18,Standard Class,Bretford CR4500 Series Slim Rectangular Table,5
Sean O'Donnell,2015-10-18,Standard Class,Eldon Fold 'N Roll Cart System,2


In [367]:
outer_join = df_1.join(df_2, how="outer").drop_duplicates()

print(f"Dimensions: {outer_join.shape}")
outer_join

Dimensions: (5, 4)


,Ship Date,Ship Mode,Product Name,Quantity
Customer Name,,,,
Brosina Hoffman,NaT,NaN,Eldon Expressions Wood and Plastic Desk Access...,7.0
Claire Gute,2016-11-11,Second Class,NaN,NaN
Darrin Van Huff,2016-06-16,Second Class,Self-Adhesive Address Labels for Typewriters b...,2.0
Sean O'Donnell,2015-10-18,Standard Class,Bretford CR4500 Series Slim Rectangular Table,5.0
Sean O'Donnell,2015-10-18,Standard Class,Eldon Fold 'N Roll Cart System,2.0


### Interpretación

- `left` conservó todos los índices de `df_1`.
- `right` conservó todos los índices de `df_2`.
- `inner` conservó únicamente los índices presentes en ambos DataFrames.
- `outer` conservó todos los índices de las dos tablas.
- Los valores `NaN` y `NaT` indican que no se encontró información correspondiente.

### Conclusión

El método `join()` permitió combinar dos DataFrames utilizando `Customer Name` 
como índice. Cada tipo de unión produjo un resultado diferente según los registros 
que debía conservar.

El `inner join` solamente mantuvo las coincidencias, mientras que el `outer join` 
conservó todos los clientes. Los métodos `left` y `right` dieron prioridad al 
DataFrame ubicado a la izquierda o a la derecha, respectivamente.

A diferencia de `merge()`, que normalmente relaciona columnas, `join()` trabaja 
principalmente con los índices de los DataFrames.

## Ejercicio 4.10: Muestreo aleatorio

El método `sample()` permite seleccionar registros aleatorios de un DataFrame.

Sus parámetros principales son:

- `n`: cantidad exacta de registros.
- `frac`: proporción del conjunto de datos.
- `replace`: permite seleccionar un mismo registro más de una vez.
- `random_state`: permite reproducir siempre la misma muestra.

In [368]:
five_sample = df.sample(n=5, random_state=42)

print(f"Sample dimensions: {five_sample.shape}")

five_sample[["Customer Name", "State", "Sales", "Profit"]]

Sample dimensions: (5, 20)


,Customer Name,State,Sales,Profit
3125,Jim Epp,Florida,563.808,21.1428
1441,Michael Moore,California,36.672,11.4600
4510,Kalyca Meade,Missouri,37.300,17.1580
39,Steve Nguyen,Texas,212.058,-15.1470
4509,Vivek Sundaresam,California,171.288,-6.4233


In [369]:
sample_10_percent = df.sample(frac=0.1, random_state=42)
sample_10_percent.head()

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3125,CA-2015-121720,2015-06-11,2015-06-12,First Class,JE-15610,Jim Epp,Corporate,United States,Lakeland,Florida,33801,South,OFF-ST-10003816,Office Supplies,Storage,Fellowes High-Stak Drawer Files,563.808,4,0.2,21.1428
1441,CA-2017-128160,2017-12-19,2017-12-24,Second Class,MM-17920,Michael Moore,Consumer,United States,San Francisco,California,94110,West,OFF-BI-10001510,Office Supplies,Binders,Deluxe Heavy-Duty Vinyl Round Ring Binder,36.672,2,0.2,11.4600
4510,CA-2016-119935,2016-11-10,2016-11-14,Standard Class,KM-16225,Kalyca Meade,Corporate,United States,Springfield,Missouri,65807,Central,FUR-FU-10001085,Furniture,Furnishings,3M Polarizing Light Filter Sleeves,37.300,2,0.0,17.1580
39,CA-2015-117415,2015-12-27,2015-12-31,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,Texas,77041,Central,FUR-CH-10004218,Furniture,Chairs,"Global Fabric Manager's Chair, Dark Gray",212.058,3,0.3,-15.1470
4509,CA-2017-167003,2017-05-23,2017-05-29,Standard Class,VS-21820,Vivek Sundaresam,Consumer,United States,Los Angeles,California,90036,West,FUR-TA-10001520,Furniture,Tables,"Lesro Sheffield Collection Coffee Table, End T...",171.288,3,0.2,-6.4233


In [370]:
sample_with_replacement = df.sample(frac=0.1, random_state=42, replace=True)
sample_with_replacement.head()

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
7270,CA-2016-123337,2016-09-16,2016-09-22,Standard Class,KD-16495,Keith Dawkins,Corporate,United States,San Jose,California,95123,West,FUR-BO-10001918,Furniture,Bookcases,"Sauder Forest Hills Library with Doors, Woodla...",273.666,2,0.15,-12.8784
860,CA-2014-133851,2014-06-09,2014-06-16,Standard Class,CM-12445,Chuck Magee,Consumer,United States,San Francisco,California,94122,West,OFF-SU-10001225,Office Supplies,Supplies,Staple remover,7.360,2,0.00,0.1472
5390,CA-2017-163874,2017-11-10,2017-11-12,Second Class,NC-18535,Nick Crebassa,Corporate,United States,Seattle,Washington,98105,West,OFF-BI-10000088,Office Supplies,Binders,GBC Imprintable Covers,26.352,3,0.20,9.5526
5191,US-2017-141943,2017-05-05,2017-05-09,Standard Class,DK-12985,Darren Koutras,Consumer,United States,San Francisco,California,94122,West,OFF-EN-10003448,Office Supplies,Envelopes,"Peel & Seel Recycled Catalog Envelopes, Brown",23.160,2,0.00,11.5800
5734,CA-2017-143574,2017-06-29,2017-07-02,First Class,DR-12880,Dan Reichenbach,Corporate,United States,Milford,Connecticut,6460,East,FUR-FU-10003976,Furniture,Furnishings,"DAX Executive Solid Wood Document Frame, Deskt...",25.160,2,0.00,8.5544


### Conclusión

El método `sample()` permitió obtener muestras aleatorias del conjunto de datos. 
El parámetro `n` seleccionó una cantidad exacta de registros, mientras que `frac` 
seleccionó una proporción del DataFrame.

Al utilizar `replace=False`, cada registro pudo aparecer solamente una vez. En 
cambio, con `replace=True`, algunos registros fueron seleccionados repetidamente. 
Finalmente, `random_state` hizo posible reproducir los mismos resultados cada vez 
que se ejecutó el notebook.

## Ejercicio 4.11: Ordenamiento con sort_values()

El método `sort_values()` permite ordenar los registros de un DataFrame utilizando 
una o varias columnas.

Por defecto, pandas ordena los valores de manera ascendente. Cuando se utilizan 
varias columnas, se aplica primero el orden de la primera y después el de las 
columnas siguientes.

In [371]:
df_sample = df[["Customer Name", "State", "Sales", "Quantity"]].sample(n=15, random_state=42)
df_sample

,Customer Name,State,Sales,Quantity
3125,Jim Epp,Florida,563.808,4
1441,Michael Moore,California,36.672,2
4510,Kalyca Meade,Missouri,37.300,2
39,Steve Nguyen,Texas,212.058,3
4509,Vivek Sundaresam,California,171.288,3
1599,Aimee Bixby,Texas,99.136,4
7363,David Smith,Pennsylvania,340.182,3
9951,Helen Wasserman,Indiana,17.380,2
500,Maribeth Yedwab,Colorado,22.620,2
4847,Grace Kelly,Oregon,71.632,11


In [372]:
sorted_sales_sample = df_sample.sort_values(by="Sales")
sorted_sales_sample

,Customer Name,State,Sales,Quantity
4568,Duane Benoit,Texas,10.476,6
9951,Helen Wasserman,Indiana,17.380,2
6220,Joy Smith,Michigan,20.860,7
500,Maribeth Yedwab,Colorado,22.620,2
1441,Michael Moore,California,36.672,2
4510,Kalyca Meade,Missouri,37.300,2
4847,Grace Kelly,Oregon,71.632,11
2758,Julie Prescott,Georgia,76.920,4
1599,Aimee Bixby,Texas,99.136,4
4509,Vivek Sundaresam,California,171.288,3


In [373]:
sorted_sales_state_sample = df_sample.sort_values(by=["State", "Sales"])
sorted_sales_state_sample

,Customer Name,State,Sales,Quantity
1441,Michael Moore,California,36.672,2
4509,Vivek Sundaresam,California,171.288,3
500,Maribeth Yedwab,Colorado,22.620,2
7729,Tanja Norvell,Florida,431.976,3
3125,Jim Epp,Florida,563.808,4
2758,Julie Prescott,Georgia,76.920,4
9951,Helen Wasserman,Indiana,17.380,2
6220,Joy Smith,Michigan,20.860,7
4510,Kalyca Meade,Missouri,37.300,2
713,Roland Schwarz,New York,307.980,2


### Interpretación

Al ordenar únicamente por `Sales`, los registros se organizaron desde la venta 
más pequeña hasta la más grande.

Al ordenar por `State` y `Sales`, pandas organizó primero los estados 
alfabéticamente. Después ordenó las ventas correspondientes a cada estado. 
El parámetro `ascending` permitió controlar si el orden era ascendente o descendente.

### Conclusión

El método `sort_values()` permitió reorganizar la muestra mediante una o varias 
columnas. El ordenamiento por ventas facilitó la identificación de los valores 
mínimos y máximos, mientras que el ordenamiento por estado y ventas permitió 
organizar los registros jerárquicamente.

Ordenar los datos facilita su interpretación y ayuda a detectar tendencias, valores 
extremos y relaciones entre diferentes categorías.

## Ejercicio 4.12: Funciones personalizadas con apply()

El método `apply()` permite ejecutar una función sobre cada elemento de una 
columna. En este ejercicio se utiliza para:

- Clasificar las ventas como bajas, medias o altas.
- Calcular la longitud del nombre de cada cliente.
- Aplicar un descuento mediante una función lambda.

In [374]:
def categorize_sales(price):
    if price < 50:
        return "Low"
    elif price < 200:
        return "Medium"
    else:
        return "High"

In [375]:
df_sample = df[["Customer Name", "State", "Sales"]].sample(n=100, random_state=42)
df_sample.head(10)

,Customer Name,State,Sales
3125,Jim Epp,Florida,563.808
1441,Michael Moore,California,36.672
4510,Kalyca Meade,Missouri,37.300
39,Steve Nguyen,Texas,212.058
4509,Vivek Sundaresam,California,171.288
1599,Aimee Bixby,Texas,99.136
7363,David Smith,Pennsylvania,340.182
9951,Helen Wasserman,Indiana,17.380
500,Maribeth Yedwab,Colorado,22.620
4847,Grace Kelly,Oregon,71.632


In [376]:
df_sample["Sales Price Category"] = df_sample["Sales"].apply(categorize_sales)
df_sample.head(10)

,Customer Name,State,Sales,Sales Price Category
3125,Jim Epp,Florida,563.808,High
1441,Michael Moore,California,36.672,Low
4510,Kalyca Meade,Missouri,37.300,Low
39,Steve Nguyen,Texas,212.058,High
4509,Vivek Sundaresam,California,171.288,Medium
1599,Aimee Bixby,Texas,99.136,Medium
7363,David Smith,Pennsylvania,340.182,High
9951,Helen Wasserman,Indiana,17.380,Low
500,Maribeth Yedwab,Colorado,22.620,Low
4847,Grace Kelly,Oregon,71.632,Medium


In [377]:
df_sample["Customer Name Lenght"] = df_sample["Customer Name"].apply(len)
df_sample.head(10)

,Customer Name,State,Sales,Sales Price Category,Customer Name Lenght
3125,Jim Epp,Florida,563.808,High,7
1441,Michael Moore,California,36.672,Low,13
4510,Kalyca Meade,Missouri,37.300,Low,12
39,Steve Nguyen,Texas,212.058,High,12
4509,Vivek Sundaresam,California,171.288,Medium,16
1599,Aimee Bixby,Texas,99.136,Medium,11
7363,David Smith,Pennsylvania,340.182,High,11
9951,Helen Wasserman,Indiana,17.380,Low,15
500,Maribeth Yedwab,Colorado,22.620,Low,15
4847,Grace Kelly,Oregon,71.632,Medium,11


In [378]:
df_sample["Discounted Price"] = df_sample["Sales"].apply(lambda price: price * 0.85 if price > 200 else price)
df_sample.head(10)

,Customer Name,State,Sales,Sales Price Category,Customer Name Lenght,Discounted Price
3125,Jim Epp,Florida,563.808,High,7,479.2368
1441,Michael Moore,California,36.672,Low,13,36.6720
4510,Kalyca Meade,Missouri,37.300,Low,12,37.3000
39,Steve Nguyen,Texas,212.058,High,12,180.2493
4509,Vivek Sundaresam,California,171.288,Medium,16,171.2880
1599,Aimee Bixby,Texas,99.136,Medium,11,99.1360
7363,David Smith,Pennsylvania,340.182,High,11,289.1547
9951,Helen Wasserman,Indiana,17.380,Low,15,17.3800
500,Maribeth Yedwab,Colorado,22.620,Low,15,22.6200
4847,Grace Kelly,Oregon,71.632,Medium,11,71.6320


### Interpretación

La función `categorize_sales()` transformó los valores numéricos de `Sales` en 
categorías de texto. Esto permite interpretar las ventas como bajas, medias o altas.

La función incorporada `len` calculó la longitud de los nombres de los clientes. 
Finalmente, una función lambda aplicó un descuento del 15 % únicamente a las 
ventas mayores que 200.

### Conclusión

El método `apply()` permitió ejecutar diferentes funciones sobre los elementos de 
las columnas del DataFrame. Se utilizó una función personalizada para clasificar las 
ventas, una función incorporada para calcular la longitud de los nombres y una 
función lambda para aplicar descuentos condicionales.

Esta herramienta proporciona flexibilidad para crear nuevas variables, transformar 
datos y aplicar reglas de negocio durante el proceso de Data Wrangling.